### DISCUSSION TITLE
Security bug - How anyone can write to the repository on behalf of any user

### DISCUSSION THREAD

**Discussion by ShakirovI** [community/community#40755] [Repositories] [answered]
### Select Topic Area

Bug

### Body

Hello)
I found how to write commits on behalf of any user without his permission. How to reproduce:
1) Open phpstorm
2) create a new project or open an existing one
3) click in phpstorm "login to github" and log in with a token or through the web interface
4) in the console, being in the project folder, change the user.name and user.email to ANY you want.
5) now any of your commits will be displayed under someone else's name.

I checked this bug - I created a project, logged in via php storm in my account using a token. Then I started a new account on github to another email. I went to the console, typed two commands
```
git config --local --add user.name "SourWolf"
git config --local --add user.email "anna.fuks31@gmail.com"
```

and now, if I make a commit from phpstrom, and now if I make a commit from phpstorm, my profile will show that the author of this commit is "SourWolf"

**Comment by supervoid0** [comment, marked as answer]
This is not a bug. This is your local git username and email. After you clone a repository that repository is cloned and the local git user is able to write in a local copy. You can modify it in whatever name you set for the local user. Multiple authors can write and commit to this local repo by changing usernames. But when you are pushing to the origin (wherever the origin is added), you have to be authenticated as that specific remote GitHub (origin) owner and you are allowing other authors' work to be added to your remote repo in GitHub. Here anyone is not writing to the repository that you own. Anyone is writing in the local copy they have cloned. Here, in this case, you are pushing the writing of Sourwolf as an authenticated owner of the remote repository.

**Reply by daniel-cues** [reply]
The issue he's describing is technically an impersonation, even though you need write access to the repo

**Reply by ShakirovI** [reply]
I understand that locally I can specify ANY name and that's fine. I also understand that I can make changes from any local name to the repository if I have a token and this is also normal.
But the fact is that if I specify someone's name, github will not only allow me to do this, but also display the profile picture of this user and a link to the current profile in the history.

**Reply by daniel-cues** [reply]
@ShakirovI As a stated in my answer, GitHub already knows this, apparently it is intended behaviour. I don’t think either that unverified commits should link to somebody’s profile and pfp, but GitHub’s answer to this is vigilant mode.

**Reply by supervoid0** [reply]
@ShakirovI ah I see, now I understand your concern.

**Comment by daniel-cues** [comment]
Hi, @ShakirovI.

I'm afraid you won't get any response from GitHub as this was already addressed a while ago.

I do agree as well this shouldn't be the case by default. However, Github's solution is "**Vigilant mode**":

![https://docs.github.com/en/authentication/managing-commit-signature-verification/displaying-verification-statuses-for-all-of-your-commits](https://docs.github.com/assets/cb-91754/images/help/commits/signature-verification-statuses.png)

When you enable _Vigilant mode_ GitHub, will show if a commit is verified or not based on the PGP keys uploaded to your profile. If you enable this, every commit not signed with any of your keys and not performed from GitHub UI will be marked as Unverified. Why unverified commits still link to the profile, I do not know, but it is what it is.

**Reply by spenserblack** [reply]
Yes, commit signatures are probably the best way to prevent impersonation :+1:

I'll add that, if you want to replicate this validation locally, you can find a user's public key by going to https://github.com/[USER].gpg. Then you can import it locally. `git log --show-signature` will show known and unkown signatures for commits.

Also, https://github.com/web-flow.gpg is the public key GitHub uses to sign commits made on the web.

**Reply by daniel-cues** [reply]
Suggesting @ShakirovI to mark as answered. Cheers!

**Comment by jfullstackdev** [comment]
do you mean, you can do that when you commit on GitHub too ? I doubt you can. You need authentication for verified commits,

like GitHub Desktop, if I did not setup the Git locally, that it is still unverified, if I will be using another username, I need authentication to do so

so it means, you can only impersonate locally that makes no sense because it is only your copy, when you push that to GitHub, then either it will not be recognized or it is unverified, 

same thing is happening when you have 2 or more users on the same computer, it will really be a big headache, either it will recognize one but will be losing all the commits of the other accounts on the same computer reflected on GitHub, this happens every time

**Reply by daniel-cues** [reply]
You can commit but it will show as unverified if you enable vigilant mode. If you don’t enable it, it’d just show as any other commit.

**Reply by jfullstackdev** [reply]
I'm not using that, without any proper setup and authentication, it will always be unverified 

![image](https://user-images.githubusercontent.com/47092464/206199002-81d39c3b-db95-4d1d-b144-056db26af2b2.png)

same as in the official documentation, it's just an added option

https://docs.github.com/en/authentication/managing-commit-signature-verification/displaying-verification-statuses-for-all-of-your-commits#about-vigilant-mode

and there is no bug mentioned there that you can do that, again, vigilante mode just increases the display `verified`, `partially verified` and `unverified`

**Reply by daniel-cues** [reply]
Unverified commits still link to the user profile and display their pfp. The whole reason verified commits exist is to be able to force verification on a whole repository and to show true identity on commits. I don’t really follow what you are trying to say. I can impersonate you in a repo of myself if you’d like

**Reply by jfullstackdev** [reply]
try it, it's better to see it, post it here 

if that is verified, it's a security risk, if that is unverified, that makes no sense, that's not me, that's the point

use both with no vigilante and vigilante mode

**Reply by jfullstackdev** [reply]
I'm already seeing it, and as expected, that makes no sense

![image](https://user-images.githubusercontent.com/47092464/206217589-b7bf3e29-8283-400e-9200-606dcfde81ac.png)

even with GitHub Desktop and without my Git setup, I will have the same thing, same thing I'm telling above

please do it with Vigilante Mode

**Reply by daniel-cues** [reply]
Here you go @jdevstatic https://github.com/daniel-cues/spoof-test

![Screenshot 2022-12-07 at 16 16 52](https://user-images.githubusercontent.com/16452665/206217513-aa891229-21bc-4c64-9f2d-b107b819e998.png)

If you had Vigilant Mode enabled It would just show an **(Unverified)** tag before the commit hash, but everything else (profile picture, username and profile link) would be the same:

![Screenshot 2022-12-07 at 16 21 12](https://user-images.githubusercontent.com/16452665/206218792-75b839ce-a5f6-4ae7-b26d-8894f8ecb1ef.png)

As I said, this has been known for ages, vigilant mode is the solution GitHub provided. A bad solution? Yes, but a solution nonetheless.

**Reply by jfullstackdev** [reply]
yup, it's the nature of Git, so why anyone would say it's a GitHub bug when their closest solution is this one,

see, as I was telling above, that does not make any sense, I will just say, it's unverified, it's not my commit, I can do the same thing on repos I have write access if I will not be setting up Git but using GitHub Desktop and I can say, that's not my commit

in short, in order to be verified, there must be some authentication that proves you are the committer and not a spoof or else that would be one of the greatest security issues

**Reply by daniel-cues** [reply]
I do believe as well that vigilant mode should let just block any unverified commits, and that it should be enabled by default. However, I'm sure that the team at @github-staff know what they're doing.

**Reply by daniel-cues** [reply]
> I'm already seeing it, and as expected, that makes no sense
> 
> ![image](https://user-images.githubusercontent.com/47092464/206217589-b7bf3e29-8283-400e-9200-606dcfde81ac.png)
> 
> even with GitHub Desktop and without my Git setup, I will have the same thing, same thing I'm telling above
> 
> please do it with Vigilante Mode

No, you are not following me, Vigilant mode is a setting each user can enable for themselves, the only difference would be the Unverified tag

**Reply by jfullstackdev** [reply]
yup, that will be a great feature if they will be blocking all unverified commits but for now, I think that's a sufficient solution too, as it is Git nature and they just did that unverified commits

**Comment by spenserblack** [comment]
Combined with the ability to access a fork's ref from the main repo, this can result in some [interesting behavior :laughing:](https://github.com/torvalds/linux/tree/5895e21f3c744ed9829e3afe9691e3eb1b1932ae)

**Comment by peppy** [comment]
As a real world example, someone is impersonating me [here](https://github.com/7ez/osuclient.ts/commits/master), but this isn't the first time it's happened. I understand I could turn on "vigilant" mode, but that's also going to mark all *my* old commits as unverified. I really do thing that something should be done about this if only as a change in display / profile linking on the github site. It can be used for abuse / harassment currently. People have used my name and triggered mass notifications to show up in my github inbox as a result.

**Comment by hakuna247** [comment]
haliunaacecgee@gmail.com
